### PR TITLE
TBranch: extend bulk io documentation

### PR DESCRIPTION
**Pull request by pcanal**

### CODE REVIEW

File: tree/tree/src/TBranch.cxx [hunk 1]
```diff
@@ -1530,10 +1529,43 @@ Int_t TBranch::GetBulkEntries(Long64_t entry, TBuffer &user_buf)
...
+/// For each entry the number of elements is the multiplication of
+///    TLeaf *leaf = dynamic_cast<TLeaf*>(branch->GetListOfLeaves()->At(0))
+///    auto len = leaf->GetLen();
+/// and the value in the BranchCount corresponding to that entry (can be obtained
+/// from branch->GetBranchCount())
+///
+/// NOTES:
+/// - This interface is meant to be used by higher-level, type-safe wrappers, not
+///   by end-users.
+/// - This only returns events
```
Comment (eguiraud) [inline tree/tree/src/TBranch.cxx:1562]: here and above, what does "This only returns events" imply? as opposed to..?

File: tree/tree/inc/TBranch.h [hunk 2]
```diff
@@ -70,9 +70,13 @@ class TBulkBranchRead {
    friend class ::TBranch;
 
 public:
+   // See TBranch::GetBulkEntries(Long64_t evt, TBuffer &user_buf);
    Int_t GetBulkEntries(Long64_t evt, TBuffer &user_buf);
+   // See TBranch::GetEntriesSerialized(Long64_t evt, TBuffer &user_buf);
    Int_t GetEntriesSerialized(Long64_t evt, TBuffer &user_buf);
+   // See TBranch::GetEntriesSerialized(Long64_t evt, TBuffer &user_buf, TBuffer *count_buf);
    Int_t GetEntriesSerialized(Long64_t evt, TBuffer &user_buf, TBuffer *count_buf);
+   // Returns true if the branch can be read through the bulk interfaces.
```
Comment (eguiraud) [inline tree/tree/inc/TBranch.h:79]:
```suggestion
   /// Return true if the branch can be read through the bulk interfaces.
```

File: tree/tree/src/TBranch.cxx [hunk 3]
```diff
@@ -1530,10 +1529,43 @@ Int_t TBranch::GetBulkEntries(Long64_t entry, TBuffer &user_buf)
    return N;
 }
 
-// TODO: Template this and the call above; only difference is the TLeaf function (ReadBasketFast vs
-// ReadBasketSerialized
+////////////////////////////////////////////////////////////////////////////////
+/// Read as many events as possible into the given buffer, using zero-copy
+/// mechanisms.
```
Comment (eguiraud) [inline tree/tree/src/TBranch.cxx:1534]: No reason not to be more explicit? :smile: 

```suggestion
/// mechanisms (e.g. no byte-swaps are performed).
```

File: tree/tree/src/TBranch.cxx [hunk 4]
```diff
@@ -1530,10 +1529,43 @@ Int_t TBranch::GetBulkEntries(Long64_t entry, TBuffer &user_buf)
...
+///
+/// Returns -1 in case of a failure.  On success, returns a (non-zero) number of
+/// events of the type held by this branch currently in the buffer.
+///
+/// On success, the caller still need to deserialize the content.  For example for
+/// a scalar branch and `N` the return value (i.e. number of entries)
+///
+///     rawdata = static_cast<char*>(buf.GetCurrent());
+///     for (std::size_t i = 0u; i < N; ++i, ++target)
+///         frombuf(rawdata, target); // `frombuf` also advances the `data` pointer
```
Comment (eguiraud) [inline tree/tree/src/TBranch.cxx:1544]:
```suggestion
///         frombuf(rawdata, target); // `frombuf` also advances the `rawdata` pointer
```

File: tree/tree/src/TBranch.cxx [hunk 5]
```diff
@@ -1530,10 +1529,43 @@ Int_t TBranch::GetBulkEntries(Long64_t entry, TBuffer &user_buf)
...
+///
+/// where T is the type stored on this branch.
```
Comment (eguiraud) [inline tree/tree/src/TBranch.cxx:1546]: there is no `T` in the snippet above

File: tree/tree/src/TBranch.cxx [hunk 6]
```diff
@@ -1530,10 +1529,43 @@ Int_t TBranch::GetBulkEntries(Long64_t entry, TBuffer &user_buf)
...
+///
+/// When `count_buf` points to a valid TBuffer and the branch has a branch count,
+/// `count_buf` will be filled (via a call to GetEntriesSerialized) with the data
```
Comment (eguiraud) [inline tree/tree/src/TBranch.cxx:1549]: to trigger doxygen links:
```suggestion
/// `count_buf` will be filled (via a call to GetEntriesSerialized()) with the data
```

File: tree/tree/src/TBranch.cxx [hunk 7]
```diff
@@ -1440,12 +1440,11 @@ Bool_t TBranch::SupportsBulkRead() const {
 /// Returns -1 in case of a failure.  On success, returns a (non-zero) number of
 /// events of the type held by this branch currently in the buffer.
 ///
-/// On success, the caller should be able to access the contents of buf as
+/// On success, the caller should be able to access the contents of buf as is with
```
Comment (eguiraud) [inline tree/tree/src/TBranch.cxx:1443]:
```suggestion
/// On success, the caller should be able to access the contents of buf as they are with
```

**Review by eguiraud — APPROVED** [review]
Thank you!

File: tree/tree/src/TBranch.cxx (hunk 1 above)
Comment (pcanal) [reply tree/tree/src/TBranch.cxx:1562]: That's a good question.  I am not sure what the original author meant ....

**Pull request merged**